runtime/src/main/java/org/corfudb/runtime/LogReplicationListener.java

```diff
@@ -0,0 +1,184 @@
+        this.corfuStore = corfuStore;
+        this.namespace = namespace;
+        this.streamTag = streamTag;
+        this.tablesOfInterest = tablesOfInterest;
+        this.bufferSize = bufferSize;
+    }
+
+    /**
+     * This is an internal method of this abstract listener and not exposed to clients.
+     *
+     * @param results is a map of stream UUID -> list of entries of this stream.
+     */
+    public final void onNext(CorfuStreamEntries results) {
+
+        // A corfu update can/may have multiple updates belonging to different streams.  This callback will return
+        // those updates as a list grouped by their Stream UUIDs.
+        // Note: there is no order guarantee within the transaction boundaries.
+
+        Set<String> tableNames =
+                results.getEntries().keySet().stream().map(schema -> schema.getTableName()).collect(Collectors.toSet());
+
+        if (tableNames.contains(REPLICATION_STATUS_TABLE)) {
+            Preconditions.checkState(results.getEntries().keySet().size() == 1,
+                "Replication Status Table Update received with other tables");
+            processReplicationStatusUpdate(results);
+            return;
+        }
+
+        // Data Updates
+        if (snapshotSyncInProgressOnSubscription.get()) {
+            // If the listener started when snapshot sync was already ongoing, ignore all data updates until it ends
+            return;
+        }
+
+        if (snapshotSyncInProgress.get()) {
+            processUpdatesInSnapshotSync(results);
+        } else {
+            processUpdatesInLogEntrySync(results);
+        }
+    }
+
+    private void processReplicationStatusUpdate(CorfuStreamEntries results) {
+        Map<TableSchema, List<CorfuStreamEntry>> entries = results.getEntries();
+
+        TableSchema replicationStatusTableSchema =
+                entries.keySet().stream().filter(key -> key.getTableName().equals(REPLICATION_STATUS_TABLE))
+                        .findFirst().get();
+
+        for (CorfuStreamEntry entry : entries.get(replicationStatusTableSchema)) {
+
+            // Ignore any update where the operation type != UPDATE
+            if (entry.getOperation() == CorfuStreamEntry.OperationType.UPDATE) {
+                ReplicationStatusVal status = (ReplicationStatusVal)entry.getPayload();
+
+                if (snapshotSyncInProgressOnSubscription.get() && status.getDataConsistent() == true) {
+                    snapshotSyncInProgressOnSubscription.set(false);
+                    corfuStore.unsubscribeListener(this);
+
+                    if (bufferSize == 0) {
+                        corfuStore.subscribeLogReplicationListener(this, namespace, streamTag,
+                                tablesOfInterest);
+                    } else if (tablesOfInterest == null) {
+                        corfuStore.subscribeLogReplicationListener(this, namespace, streamTag, bufferSize);
+                    } else {
```

Comment:
I'm a bit concerned with subscribe without tables of interest pattern in this use case (using stream tag to fetch), since I noticed following error during testing, for example

1. Source side has 2 tables opened: Table1[tagA, groupA], Table2[tagA, groupB], where groupA is targeting Sink1, groupB is targeting Sink2
2. RegistryTable is replicating to all Sinks, so Sink1 and Sink2 have all the registry table entries for Table1 and Table2. but Sink1 could have not opened (or will never open) Table2, and vice versa
3. Then if we use tagA to subscribe listener, we will get IllegalArgumentException when creating stream tasks, because we can get a full list of tagA table but `getTable` is only checking in-memory tableMap...

I thought we might need to always ask them to provide the tables of interest? or catch this "table hasn't been opened" IllegalArgumentException in `LRStreamingTask`?

Reply:
Good catch. Let's discuss this and we should have a test for this case (negative testing)